// @logto/node/lib/src/utils/cookie-storage.js
import { PromiseQueue } from './promise-queue.js';
import { wrapSession, unwrapSession } from './session.js';

/**
 * A storage that persists data in cookies with encryption.
 */
class CookieStorage {
    get cookieOptions() {
        return Object.freeze({
            httpOnly: true,
            path: '/',
            sameSite: 'lax',
            secure: this.config.isSecure ?? false,
            maxAge: 14 * 24 * 3600, // 14 days
        });
    }
    get cookieKey() {
        return this.config.cookieKey ?? 'logtoCookies';
    }
    get data() {
        return this.sessionData;
    }
    constructor(config) {
        this.config = config;
        this.sessionData = {};
        this.saveQueue = new PromiseQueue();
        if (!config.sessionWrapper && !config.encryptionKey) {
            throw new TypeError('Either `sessionWrapper` or `encryptionKey` must be provided for `CookieStorage`');
        }
        this.sessionWrapper = config.sessionWrapper ?? {
            wrap: wrapSession,
            unwrap: unwrapSession,
        };
    }
    async init() {
        const { encryptionKey = '' } = this.config;
        this.sessionData = await this.sessionWrapper.unwrap((await this.config.getCookie(this.cookieKey)) ?? '', encryptionKey);
    }
    async getItem(key) {
        return this.sessionData[key] ?? null;
    }
    async setItem(key, value) {
        this.sessionData[key] = value;
        await this.save();
    }
    async removeItem(key) {
        // eslint-disable-next-line @silverhand/fp/no-delete, @typescript-eslint/no-dynamic-delete
        delete this.sessionData[key];
        await this.save();
    }
    async destroy() {
        this.sessionData = {};
        await this.save();
    }
    async save() {
        return this.saveQueue.enqueue(async () => this.write());
    }
    async write(data = this.sessionData) {
        const { encryptionKey = '' } = this.config;
        const value = await this.sessionWrapper.wrap(data, encryptionKey);
        await this.config.setCookie(this.cookieKey, value, this.cookieOptions);
    }
}

export { CookieStorage };
