// @logto/node/lib/src/utils/session.js
async function getKeyFromPassword(password, crypto) {
    const encoder = new TextEncoder();
    const data = encoder.encode(password);
    const hash = await crypto.subtle.digest('SHA-256', data);
    // Convert the hash to a hex string
    return Array.from(new Uint8Array(hash))
        .map((byte) => byte.toString(16).padStart(2, '0'))
        .join('');
}
async function encrypt(text, password, crypto) {
    const iv = crypto.getRandomValues(new Uint8Array(12));
    const encodedPlaintext = new TextEncoder().encode(text);
    const secretKey = await crypto.subtle.importKey('raw', Buffer.from(await getKeyFromPassword(password, crypto), 'hex'), {
        name: 'AES-GCM',
        length: 256,
    }, true, ['encrypt', 'decrypt']);
    const ciphertext = await crypto.subtle.encrypt({
        name: 'AES-GCM',
        iv,
    }, secretKey, encodedPlaintext);
    return {
        ciphertext: Buffer.from(ciphertext).toString('base64'),
        iv: Buffer.from(iv).toString('base64'),
    };
}
async function decrypt(ciphertext, iv, password, crypto) {
    const secretKey = await crypto.subtle.importKey('raw', Buffer.from(await getKeyFromPassword(password, crypto), 'hex'), {
        name: 'AES-GCM',
        length: 256,
    }, true, ['encrypt', 'decrypt']);
    const cleartext = await crypto.subtle.decrypt({
        name: 'AES-GCM',
        iv: Buffer.from(iv, 'base64'),
    }, secretKey, Buffer.from(ciphertext, 'base64'));
    return new TextDecoder().decode(cleartext);
}
const unwrapSession = async (cookie, secret) => {
    try {
        const [ciphertext, iv] = cookie.split('.');
        if (!ciphertext || !iv) {
            return {};
        }
        const decrypted = await decrypt(ciphertext, iv, secret, crypto);
        // eslint-disable-next-line no-restricted-syntax
        return JSON.parse(decrypted);
    }
    catch {
        // Ignore invalid session
    }
    return {};
};
const wrapSession = async (session, secret) => {
    const { ciphertext, iv } = await encrypt(JSON.stringify(session), secret, crypto);
    return `${ciphertext}.${iv}`;
};

export { unwrapSession, wrapSession };
