// @logto/express/lib/index.js
import NodeClient from '@logto/node';
export { LogtoClientError, LogtoError, LogtoRequestError, OidcError, PersistKey, Prompt, ReservedResource, ReservedScope, UserScope, buildOrganizationUrn, getOrganizationIdFromUrn, organizationUrnPrefix } from '@logto/node';
import { Router } from 'express';
import { LogtoExpressError } from './errors.js';
import ExpressStorage from './storage.js';

const createNodeClient = (request, response, config) => {
    // We assume that `session` is configured in the express app, but need to check it there.
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    if (!request.session) {
        throw new LogtoExpressError('session_not_configured');
    }
    const storage = new ExpressStorage(request);
    return new NodeClient(config, {
        storage,
        navigate: (url) => {
            response.redirect(url);
        },
    });
};
const handleAuthRoutes = (config) => {
    // eslint-disable-next-line new-cap
    const router = Router();
    const prefix = config.authRoutesPrefix ?? 'logto';
    router.use(`/${prefix}/:action`, async (request, response) => {
        const { action } = request.params;
        const nodeClient = createNodeClient(request, response, config);
        switch (action) {
            case 'sign-in': {
                await nodeClient.signIn({
                    ...config.signInOptions,
                    redirectUri: `${config.baseUrl}/${prefix}/sign-in-callback`,
                });
                break;
            }
            case 'sign-up': {
                await nodeClient.signIn({
                    ...config.signInOptions,
                    redirectUri: `${config.baseUrl}/${prefix}/sign-in-callback`,
                    firstScreen: 'register',
                });
                break;
            }
            case 'sign-in-callback': {
                if (request.url) {
                    await nodeClient.handleSignInCallback(`${config.baseUrl}${request.originalUrl}`);
                    response.redirect(config.baseUrl);
                }
                break;
            }
            case 'sign-out': {
                await nodeClient.signOut(config.baseUrl);
                break;
            }
            default: {
                response.status(404).end();
            }
        }
    });
    return router;
};
const withLogto = (config) => async (request, response, next) => {
    try {
        const client = createNodeClient(request, response, config);
        const user = await client.getContext({
            getAccessToken: config.getAccessToken,
            resource: config.resource,
            fetchUserInfo: config.fetchUserInfo,
            getOrganizationToken: config.getOrganizationToken,
        });
        // eslint-disable-next-line @silverhand/fp/no-mutating-methods
        Object.defineProperty(request, 'user', { enumerable: true, get: () => user });
        next();
    }
    catch (error) {
        next(error);
    }
};

export { handleAuthRoutes, withLogto };
