// @logto/node/lib/src/client.js
import BaseClient from '@logto/client';
export { LogtoClientError, LogtoError, LogtoRequestError, OidcError, Prompt, ReservedScope, UserScope } from '@logto/client';
import { trySafe, conditional } from '@silverhand/essentials';

class LogtoNodeBaseClient extends BaseClient {
    constructor() {
        super(...arguments);
        this.getContext = async ({ getAccessToken, resource, fetchUserInfo, getOrganizationToken, organizationId, } = {}) => {
            const isAuthenticated = await this.isAuthenticated();
            if (!isAuthenticated) {
                return {
                    isAuthenticated,
                };
            }
            const claims = await this.getIdTokenClaims();
            const { accessToken, accessTokenClaims } = getAccessToken
                ? {
                    accessToken: await trySafe(async () => this.getAccessToken(resource, organizationId)),
                    accessTokenClaims: await trySafe(async () => this.getAccessTokenClaims(resource)),
                }
                : { accessToken: undefined, accessTokenClaims: undefined };
            if (getAccessToken && !accessToken) {
                // Failed to get access token, the user is not authenticated
                return {
                    isAuthenticated: false,
                };
            }
            const organizationTokens = conditional(getOrganizationToken &&
                claims.organizations &&
                Object.fromEntries(await Promise.all(claims.organizations.map(async (organizationId) => [
                    organizationId,
                    await this.getOrganizationToken(organizationId),
                ]))));
            return {
                isAuthenticated,
                claims,
                userInfo: conditional(fetchUserInfo && (await this.fetchUserInfo())),
                ...conditional(getAccessToken && { accessToken, scopes: accessTokenClaims?.scope?.split(' ') }),
                organizationTokens,
            };
        };
    }
}

export { LogtoNodeBaseClient as default };
