// @logto/node/lib/src/index.js
import { createRequester } from '@logto/client';
export { LogtoClientError, LogtoError, LogtoRequestError, OidcError, PersistKey, Prompt, ReservedResource, ReservedScope, StandardLogtoClient, UserScope, buildOrganizationUrn, getOrganizationIdFromUrn, organizationUrnPrefix } from '@logto/client';
import { generateCodeChallenge, generateCodeVerifier, generateState } from '../edge/generators.js';
import LogtoNodeBaseClient from './client.js';
export { unwrapSession, wrapSession } from './utils/session.js';
export { CookieStorage } from './utils/cookie-storage.js';

class LogtoClient extends LogtoNodeBaseClient {
    constructor(config, adapter, buildJwtVerifier) {
        super(config, {
            requester: createRequester(config.appSecret
                ? async (...args) => {
                    const [input, init] = args;
                    return fetch(input, {
                        ...init,
                        headers: {
                            Authorization: `Basic ${Buffer.from(`${config.appId}:${config.appSecret}`, 'utf8').toString('base64')}`,
                            ...init?.headers,
                        },
                    });
                }
                : fetch),
            generateCodeChallenge,
            generateCodeVerifier,
            generateState,
            ...adapter,
        }, buildJwtVerifier);
    }
}

export { LogtoClient as default };
